crates/turborepo-lib/src/run/scope/change_detector.rs

```diff
@@ -0,0 +1,177 @@
+            self.get_changes_from_lockfile(&changed_files, from_ref)?;
+
+        if !full_changes {
+            changed_pkgs.extend(lockfile_changes);
+        } else {
+            return Ok(self
+                .pkg_graph
+                .workspaces()
+                .map(|(n, _)| n.to_string())
+                .collect());
+        }
+        Ok(changed_pkgs)
+    }
+}
+
+impl<'a> SCMChangeDetector<'a> {
+    const DEFAULT_GLOBAL_DEPS: [&'static str; 2] = ["package.json", "turbo.json"];
+
+    pub fn new(
+        turbo_root: &'a AbsoluteSystemPath,
+
+        scm: &'a SCM,
+        pkg_graph: &'a PackageGraph,
+        global_deps: Vec<String>,
+        ignore_patterns: Vec<String>,
+    ) -> Self {
+        Self {
+            turbo_root,
+            scm,
+            pkg_graph,
+            global_deps,
+            ignore_patterns,
+        }
+    }
+
+    fn repo_global_file_has_changed(
+        &self,
+        default_global_deps: &[&str],
+        changed_files: &HashSet<AnchoredSystemPathBuf>,
+    ) -> Result<bool, turborepo_scm::Error> {
+        let global_deps = self.global_deps.iter().map(|s| s.as_str());
+        let filters = global_deps.chain(default_global_deps.iter().copied());
+        let matcher = wax::any(filters).unwrap();
+        Ok(changed_files.iter().any(|f| matcher.is_match(f.as_path())))
+    }
+
+    fn filter_ignored_files<'b>(
+        &self,
+        changed_files: impl Iterator<Item = &'b AnchoredSystemPathBuf> + 'b,
+    ) -> Result<HashSet<&'b AnchoredSystemPathBuf>, turborepo_scm::Error> {
+        let matcher = wax::any(self.ignore_patterns.iter().map(|s| s.as_str())).unwrap();
+        Ok(changed_files
+            .filter(move |f| !matcher.is_match(f.as_path()))
+            .collect())
+    }
+
+    // note: this could probably be optimized by using a hashmap of package paths
+    fn get_changed_packages<'b>(
+        &self,
+        files: impl Iterator<Item = &'b AnchoredSystemPathBuf>,
+        graph: &PackageGraph,
+    ) -> Result<HashSet<String>, turborepo_scm::Error> {
+        let mut changed_packages = HashSet::new();
+        for file in files {
+            let mut found = false;
+            for (name, entry) in graph.workspaces() {
+                if name == &WorkspaceName::Root {
+                    continue;
+                }
+                let package_path = entry.package_json_path();
+                if Self::is_file_in_package(file, package_path) {
+                    changed_packages.insert(name.to_string());
+                    found = true;
+                    break;
+                }
+            }
+            if !found {
+                // if the file is not in any package, it must be in the root package
+                changed_packages.insert(ROOT_PKG_NAME.to_string());
+            }
+        }
+
+        Ok(changed_packages)
+    }
+
+    fn is_file_in_package(file: &AnchoredSystemPath, package_path: &AnchoredSystemPath) -> bool {
```

Comment:
We're passing in the path to a workspace's `package.json` so if there's a changed `packages/foo/index.js` then this will fail since we're checking component equality against `packages/foo/package.json`.

Reply:
The other function seems to do essentially the same thing.

Reply:
Nice catch, I think at the time I thought that was the dir, not the file itself.